src/modules/leaderboard.ts

```diff
@@ -0,0 +1,182 @@
+import * as Discord from "discord.js";
+import * as Builders from "@discordjs/builders";
+
+import { CommandDescriptor } from "../bot.d";
+
+import * as utils from "./utils";
+
+const MAX_PEOPLE = 50;
+
+type UsersCharacterCount = Discord.Collection<Discord.Snowflake, number>;
+
+export async function getUsersCharacterCount(
+	guild: Discord.Guild,
+	onlyCountAfter?: Date
+): Promise<[UsersCharacterCount, number, number, Discord.Channel[]]> {
+	// TODO: checkpoint
+
+	const chars: UsersCharacterCount = new Discord.Collection();
+	const failed: (Discord.TextChannel | Discord.ThreadChannel)[] = [];
+	let msgCount = 0;
+
+	const channels = Array.from(
+		(await guild.channels.fetch()).filter(
+			(channel) => channel.isText() || channel.isThread()
+		)
+	).map((o) => o[1]) as (Discord.TextChannel | Discord.ThreadChannel)[];
+	const promises = channels.map(async (channel) => {
+		const messages = await channel.messages.fetch();
+		for (const [_id, msg] of messages) {
+			if (!msg.deleted && msg.author && !msg.author.bot) {
+				if (
+					onlyCountAfter === undefined ||
+					(msg.editedAt ?? msg.createdAt) >= onlyCountAfter
+				) {
+					chars.set(
+						msg.author.id,
+						(chars.get(msg.author.id) ?? 0) + msg.content.length
+					);
+					msgCount++;
+				} else {
+					// ? FIXME: TODO: !!! SHOULD A BREAK BE HERE?
+					// * would really help efficiency but I don't know if messages are sorted
+					// * either way recently edited messages wouldn't be counted
+				}
```

Comment:
Pretty sure messages are sorted, but you can check by printing them.
Maybe if they're cached they won't be in order, but I find that hard to believe.